client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/RemoteCacheImpl.java

```diff
@@ -522,11 +527,156 @@ protected void set(K key, V value) {
    }
 
    @Override
-   public Set<K> keySet() {
-      assertRemoteCacheManagerIsStarted();
-      // Use default scope
-      BulkGetKeysOperation<K> op = operationsFactory.newBulkGetKeysOperation(0);
-      return Collections.unmodifiableSet(op.execute());
+   public CloseableIteratorSet<K> keySet() {
+      return new KeySet();
+   }
+
+   @Override
+   public CloseableIteratorSet<Entry<K, V>> entrySet() {
+      return new EntrySet();
+   }
+
+   @Override
+   public CloseableIteratorCollection<V> values() {
+      return new ValuesCollection();
+   }
+
+   private class KeySet extends AbstractCollection<K> implements CloseableIteratorSet<K> {
+
+      @Override
+      public CloseableIterator<K> iterator() {
+         return new RemovableCloseableIterator<>(
+               new CloseableIteratorMapper<>(retrieveEntries(null, batchSize), e -> (K) e.getKey()),
+               RemoteCacheImpl.this::remove);
+      }
+
+      @Override
+      public CloseableSpliterator<K> spliterator() {
+         return Closeables.spliterator(iterator(), Long.MAX_VALUE, Spliterator.NONNULL | Spliterator.CONCURRENT |
+               Spliterator.DISTINCT);
+      }
+
+      @Override
+      public int size() {
+         return RemoteCacheImpl.this.size();
+      }
+
+      @Override
+      public boolean contains(Object o) {
+         return RemoteCacheImpl.this.containsKey(o);
+      }
+
+      @Override
+      public boolean remove(Object o) {
+         return RemoteCacheImpl.this.remove(o) != null;
+      }
+
+      @Override
+      public boolean removeAll(Collection<?> c) {
+         boolean removedSomething = false;
+         for (Object removeMe : c) {
+            removedSomething |= remove(removeMe);
+         }
+         return removedSomething;
+      }
+   }
+
+   private CloseableIterator<Entry<K, V>> castIterator(CloseableIterator iterator) {
+      return iterator;
+   }
+
+   private boolean removeEntry(Map.Entry<K, V> entry) {
+      K key = entry.getKey();
+      VersionedValue<V> versionedValue = RemoteCacheImpl.this.getVersioned(key);
```

Comment:
If the entries retrieved through `retrieveEntries` don't contain that info (that would be no surprise), the `Converter` on server side has access to the `Metadata`, so it could add those to the returned entries.